File: scripts/memory-lane-commit.mjs
#!/usr/bin/env node
// scripts/memory-lane-commit.mjs
// Appends commit metadata to active lane Docker entity (PostCommit hook)

import { readFileSync, writeFileSync, renameSync } from 'fs';
import { join } from 'path';
import { execSync } from 'child_process';

const GATEWAY = process.env.MCP_GATEWAY_URL ?? 'http://127.0.0.1:3100';
const PROJECT_ROOT = (() => {
  try { return execSync('git rev-parse --show-toplevel', { encoding: 'utf8' }).trim(); } catch { return process.cwd(); }
})();

function readJson(path) {
  try { return JSON.parse(readFileSync(path, 'utf8')); } catch { return null; }
}

function writeJson(path, data) {
  const tmp = `${path}.tmp`;
  try {
    writeFileSync(tmp, JSON.stringify(data, null, 2) + '\n', 'utf8');
    renameSync(tmp, path);
  } catch { /* ignore */ }
}

async function memoryCall(toolName, args, timeoutMs = 1800) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(`${GATEWAY}/memory/tools/call`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name: toolName, arguments: args }),
      signal: ctrl.signal,
    });
    clearTimeout(timer);
    return res.ok;
  } catch {
    clearTimeout(timer);
    return false;
  }
}

async function checkDockerHealth() {
  try {
    const ctrl = new AbortController();
    setTimeout(() => ctrl.abort(), 1000);
    const r = await fetch(`${GATEWAY}/health`, { signal: ctrl.signal });
    return r.ok;
  } catch { return false; }
}

// Build updated fallback: prepend new commit, keep last 3
function buildUpdatedFallback(currentFallback, newCommitLine) {
  const existing = (currentFallback ?? '').split(' | ').map(s => s.trim()).filter(Boolean);
  const updated = [newCommitLine, ...existing].slice(0, 3);
  return updated.join(' | ');
}

async function main() {
  const activeBranchPath = join(PROJECT_ROOT, 'config', 'active-branch.json');
  const config = readJson(activeBranchPath);

  // Get current git branch
  let currentGitBranch = 'unknown';
  try { currentGitBranch = execSync('git rev-parse --abbrev-ref HEAD', { encoding: 'utf8', cwd: PROJECT_ROOT }).trim(); } catch { /* ignore */ }

  // Only run if branch matches configured branch
  if (!config || config.branch !== currentGitBranch) {
    process.exit(0);
  }

  const entity = config.entity ?? 'electrical-website-state';

  // Get commit info
  let commitInfo = 'unknown commit';
  try { commitInfo = execSync('git log -1 --format="%h %s"', { encoding: 'utf8', cwd: PROJECT_ROOT }).trim(); } catch { /* ignore */ }

  // Get changed files (max 10)
  let changedFiles = [];
  try {
    const filesOutput = execSync(
      'git diff HEAD~1 --name-only 2>/dev/null || git show --name-only --format="" HEAD',
      { encoding: 'utf8', cwd: PROJECT_ROOT, shell: '/bin/bash' }
    ).trim();
    changedFiles = filesOutput.split('\n').filter(Boolean).slice(0, 10);
  } catch { /* ignore */ }

  // Build observation string
  const firstThree = changedFiles.slice(0, 3).join(', ');
  const extra = changedFiles.length > 3 ? ` (+${changedFiles.length - 3} more)` : '';
  const filesStr = firstThree ? `files: ${firstThree}${extra}` : 'files: (none)';
  const now = new Date().toISOString();
  const observation = `commit: ${commitInfo} — ${filesStr} | at: ${now}`;

  // Docker health check (1s — hard cutoff)
  const dockerOnline = await checkDockerHealth();

  if (dockerOnline) {
    await memoryCall('add_observations', {
      observations: [{
        entityName: entity,
        contents: [observation],
      }],
    }, 1800);
  }

  // Update active-branch.json — updatedAt + fallback only
  const newFallback = buildUpdatedFallback(config.fallback, commitInfo);
  const updatedConfig = {
    branch: config.branch,
    entity: config.entity,
    fallback: newFallback,
    updatedAt: now,
  };
  writeJson(activeBranchPath, updatedConfig);

  process.exit(0);
}

main().catch(err => {
  // Never block git commit
  process.stderr.write(`[lane:commit] Warning: ${err?.message ?? String(err)}\n`);
  process.exit(0);
});

File: scripts/memory_lane_commit.py
#!/usr/bin/env python3
"""
Post-commit hook — updates active-branch.json fallback field with latest commit.
"""
from __future__ import annotations
import json
import subprocess
import sys
import urllib.request
import urllib.error
from pathlib import Path
from datetime import datetime, timezone

PROJECT_ROOT = Path(subprocess.run(
    ['git', 'rev-parse', '--show-toplevel'],
    capture_output=True, text=True
).stdout.strip())
GATEWAY = 'http://127.0.0.1:3100'


def get_git_branch() -> str:
    result = subprocess.run(['git', 'rev-parse', '--abbrev-ref', 'HEAD'],
                            capture_output=True, text=True, cwd=PROJECT_ROOT)
    return result.stdout.strip() if result.returncode == 0 else 'unknown'


def get_latest_commit() -> str:
    result = subprocess.run(['git', 'log', '-1', '--format=%h %s'],
                            capture_output=True, text=True, cwd=PROJECT_ROOT)
    return result.stdout.strip() if result.returncode == 0 else 'unknown commit'


def get_changed_files() -> list[str]:
    result = subprocess.run(
        ['git', 'diff', 'HEAD~1', '--name-only'],
        capture_output=True, text=True, cwd=PROJECT_ROOT
    )
    if result.returncode != 0:
        result = subprocess.run(
            ['git', 'show', '--name-only', '--format=', 'HEAD'],
            capture_output=True, text=True, cwd=PROJECT_ROOT
        )
    return [f for f in result.stdout.strip().splitlines() if f][:10]


def read_json(path: Path) -> dict:
    try:
        return json.loads(path.read_text(encoding='utf-8'))
    except Exception:
        return {}


def write_json(path: Path, data: dict) -> None:
    try:
        path.write_text(json.dumps(data, indent=2) + '\n', encoding='utf-8')
    except Exception:
        pass


def check_docker_health(timeout: float = 1.0) -> bool:
    try:
        req = urllib.request.Request(f'{GATEWAY}/health')
        with urllib.request.urlopen(req, timeout=timeout):
            return True
    except Exception:
        return False


def mcp_call(tool_name: str, args: dict, timeout: float = 1.8) -> bool:
    try:
        payload = json.dumps({'name': tool_name, 'arguments': args}).encode()
        req = urllib.request.Request(
            f'{GATEWAY}/memory/tools/call',
            data=payload,
            headers={'Content-Type': 'application/json'},
            method='POST',
        )
        with urllib.request.urlopen(req, timeout=timeout):
            return True
    except Exception:
        return False


def main() -> None:
    active_branch_path = PROJECT_ROOT / 'config' / 'active-branch.json'
    config = read_json(active_branch_path)
    if not config:
        return

    entity = config.get('entity', '')
    config_branch = config.get('branch', '')
    current_branch = get_git_branch()

    # Only run if branch matches
    if config_branch != current_branch:
        return

    commit_info = get_latest_commit()
    changed_files = get_changed_files()
    first_three = ', '.join(changed_files[:3])
    extra = f' (+{len(changed_files) - 3} more)' if len(changed_files) > 3 else ''
    files_str = f'files: {first_three}{extra}' if first_three else 'files: (none)'
    now = datetime.now(timezone.utc).isoformat()
    observation = f'commit: {commit_info} — {files_str} | at: {now}'

    # Update Docker
    docker_online = check_docker_health()
    if docker_online and entity:
        mcp_call('add_observations', {
            'observations': [{'entityName': entity, 'contents': [observation]}]
        })

    # Update fallback: prepend new commit, keep last 3
    existing_fallback = config.get('fallback', '')
    existing_entries = [e.strip() for e in existing_fallback.split(' | ') if e.strip()]
    commit_short = commit_info[:60]
    new_entries = [commit_short] + existing_entries
    new_fallback = ' | '.join(new_entries[:3])

    config['fallback'] = new_fallback
    config['updatedAt'] = now
    write_json(active_branch_path, config)


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f'[lane:commit] Warning: {e}', file=sys.stderr)
    sys.exit(0)
